### PR TITLE
fix server crash when accessing a malformed URI

### PR DESCRIPTION
**Pull request by ivanhofer**
Fix for #5090

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by ivanhofer** [comment]
I'm not sure how to fix the test. I would need some help there.

**Comment by benmccann** [comment]
@ivanhofer I'm not sure if you've noticed, but the tests appear to be failing with this PR

**Comment by benmccann** [comment]
Ah, sorry, I didn't see your comment.

It seems like the failure is coming from Vite. Probably the best course of action would be to create a small project that reproduces this using only Vite and not SvelteKit and file a bug there. See https://github.com/sveltejs/kit#bug-reporting for more details on that

It's not clear to me that this PR actually fixes anything. Is there a reason you think it might? Or should we go ahead and close this? At the very least I would mark it as a draft to make it clearer that it's not in a state to be merged

**Comment by ivanhofer** [comment]
I don't think this has anything to do with `vite`. It is an error that SvelteKit does not handle yet. I think this might only occur when using the node adapter. I just added this info to the issue description.

The fix works. We are using it since a few weeks in two of our applications.
Also when testing it locally I get the expected outcome : a 400 error.
The tests in this CI process deliver a 500 error and I don't really get why. That's why I'm asking for help because I could not figure out what happens inside this test setup and I don't want to change something that could affect other tests.

**Comment by ivanhofer** [comment]
Ah, now I see that there are two different "test" calls "test:dev" and "test:prod". So there is actually a bug in SvelteKit and in vite. I'll also open an issue there.

Still there is an error when running "test:prod" with the `javaScriptEnabled` option, that does not come from vite. It comes from how the tests are setup.

**Comment by benmccann** [comment]
I took a look at filing a Vite issue this morning, but I think Vite's fine actually because the server doesn't crash entirely like the SvelteKit server does. We just need to update the test in this PR

**Review by benmccann — APPROVED** [review]

**Comment by ivanhofer** [comment]
Smart way to handle the test 👍
Thanks for your investigation and help!

File: packages/kit/src/runtime/server/index.js [hunk 1]
```diff
@@ -41,7 +41,12 @@ export async function respond(request, options, state) {
 		}
 	}
 
-	let decoded = decodeURI(url.pathname);
+	let decoded;
+	try {
+		decoded = decodeURI(url.pathname);
+	} catch {
+		return new Response(undefined, { status: 400 });
```
[the remark this answers is not in the document]
Comment (ivanhofer) [reply packages/kit/src/runtime/server/index.js:48]: logging the url makes sense. I took inspiration from [line 59](https://github.com/ivanhofer/kit/blob/fix-server-crash-on-malformed-uri/packages/kit/src/runtime/server/index.js#L59) of this file that also returns `undefined`.
Should we also add a meaningful error message there?

[the remark this answers is not in the document]
Comment (benmccann) [reply packages/kit/src/runtime/server/index.js:48]: Sounds reasonable to me

[the remark this answers is not in the document]
Comment (ivanhofer) [reply packages/kit/src/runtime/server/index.js]: I have added an error message for the malformed URI.

Regarding the 404 error below:
I'm not entirely shure when this error occurs. Should We just go with: `Could not find route '${decoded}'`?

[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/index.js]: I think a simple 'Not found' is probably fine — added that.

Is it normal to include the URL in an error response? Feels redundant given that it's just echoing the input you gave it. Any error logging the developer has set up will already include the URL, so I'm not sure what benefit there is

**Review by Rich-Harris — APPROVED** [review]

**Comment by Rich-Harris** [comment]
thanks!

**Pull request merged**